tls/bignum.c

```diff
@@ -1,1009 +1,865 @@
+/**
+ * Import X from unsigned binary data.
+ * The bytes are read in reverse order and stored as big endian.
  */
-int ttls_mpi_write_string(const ttls_mpi *X, int radix,
-				  char *buf, size_t buflen, size_t *olen)
+int
+ttls_mpi_read_binary(TlsMpi *X, const unsigned char *buf, size_t buflen)
 {
-	int ret = 0;
-	size_t n;
-	char *p;
-	ttls_mpi T;
+	size_t i = buflen, l = 0, j;
+	size_t const limbs = CHARS_TO_LIMBS(buflen);
 
-	if (radix < 2 || radix > 16)
-		return(TTLS_ERR_MPI_BAD_INPUT_DATA);
+	if (X->limbs < limbs) {
+		unsigned long *p;
 
-	n = ttls_mpi_bitlen(X);
-	if (radix >=  4) n >>= 1;
-	if (radix >= 16) n >>= 1;
-	/*
-	 * Round up the buffer length to an even value to ensure that there is
-	 * enough room for hexadecimal values that can be represented in an odd
-	 * number of digits.
-	 */
-	n += 3 + ((n + 1) & 1);
+		if (!(p = kmalloc(limbs * CIL, GFP_ATOMIC)))
+			return -ENOMEM;
+		kfree(X->p);
+		X->p = p;
+		X->limbs = limbs;
+	}
 
-	if (buflen < n)
-	{
-		*olen = n;
-		return(TTLS_ERR_MPI_BUFFER_TOO_SMALL);
+	X->s = 1;
+	while (i >= CIL) {
+		i -= CIL;
+		X->p[l] = cpu_to_be64(*(long *)(buf + i));
+		++l;
+	}
+	if (i) {
+		/* Read last, probably incomplete, limb if any. */
+		X->p[l] = 0;
+		for (j = 0; i > 0; i--, j += 8)
+			X->p[l] |= ((unsigned long)buf[i - 1]) << j;
 	}
 
-	p = buf;
-	ttls_mpi_init(&T);
+	mpi_fixup_used(X, limbs);
 
-	if (X->s == -1)
-		*p++ = '-';
+	return 0;
+}
 
-	if (radix == 16)
-	{
-		int c;
-		size_t i, j, k;
+/**
+ * Export X into unsigned binary data, big endian.
+ * Always fills the whole buffer, which will start with zeros if the number
+ * is smaller.
+ */
+int
+ttls_mpi_write_binary(const TlsMpi *X, unsigned char *buf, size_t buflen)
+{
+	size_t i = buflen, j, n;
 
-		for (i = X->n, k = 0; i > 0; i--)
-		{
-			for (j = ciL; j > 0; j--)
-			{
-				c = (X->p[i - 1] >> ((j - 1) << 3)) & 0xFF;
+	n = ttls_mpi_size(X);
 
-				if (c == 0 && k == 0 && (i + j) != 2)
-		continue;
+	if (buflen < n)
+		return -ENOSPC;
 
-				*(p++) = "0123456789ABCDEF" [c / 16];
-				*(p++) = "0123456789ABCDEF" [c % 16];
-				k = 1;
-			}
-		}
-	}
-	else
-	{
-		TTLS_MPI_CHK(ttls_mpi_copy(&T, X));
+	/* TODO #1064 use cpu_to_be64() for bytes inverted write. */
+	for (j = 0; n > 0; i--, j++, n--)
+		buf[i - 1] = (unsigned char)(X->p[j / CIL] >> ((j % CIL) << 3));
+	memset(buf, 0, i);
 
-		if (T.s == -1)
-			T.s = 1;
+	return 0;
+}
 
-		TTLS_MPI_CHK(mpi_write_hlp(&T, radix, &p));
-	}
+/**
+ * Fill X with @size bytes of random.
+ *
+ * Use a temporary bytes representation to make sure the result is the same
+ * regardless of the platform endianness (useful when f_rng is actually
+ * deterministic, eg for tests).
+ */
+int
+ttls_mpi_fill_random(TlsMpi *X, size_t size)
+{
+	size_t limbs = CHARS_TO_LIMBS(size);
 
-	*p++ = '\0';
-	*olen = p - buf;
+	if (WARN_ON_ONCE(size > TTLS_MPI_MAX_SIZE))
+		return -EINVAL;
 
-cleanup:
+	if (__mpi_realloc(X, limbs, 0))
+		return -ENOMEM;
 
-	ttls_mpi_free(&T);
+	ttls_rnd(X->p, size);
```

Comment:
This will fill `X->p` by `size` bytes of random data. However, remaining bytes in the last limb will be kept as they were after memory allocation. So it's possible to have random number larger than expected.